autoPyTorch/api/results_visualizer.py

```diff
@@ -0,0 +1,273 @@
+                         f'but got xscale={plot_setting_params.xscale}, yscale={plot_setting_params.yscale}')
+
+    n_evals, runtime_lb, runtime_ub = results.size, cum_times[0], cum_times[-1]
+
+    if plot_setting_params.xscale == 'log':
+        # Take the even time interval in the log scale and revert
+        check_points = np.exp(np.linspace(np.log(runtime_lb), np.log(runtime_ub), plot_setting_params.n_points))
+    else:
+        check_points = np.linspace(runtime_lb, runtime_ub, plot_setting_params.n_points)
+
+    # The worst possible value is always at the head
+    perf_by_time_step = np.full_like(check_points, worst_val)
+    cur = 0
+
+    for i, check_point in enumerate(check_points):
+        while cur < n_evals and cum_times[cur] <= check_point:
+            # Guarantee that cum_times[cur] > check_point
+            # ==> cum_times[cur - 1] <= check_point
+            cur += 1
+        if cur:  # filter cur - 1 == -1
+            # results[cur - 1] was obtained before or at the checkpoint
+            # ==> The best performance up to this checkpoint
+            perf_by_time_step[i] = results[cur - 1]
+
+    if plot_setting_params.yscale == 'log' and np.any(perf_by_time_step < 0):
+        raise ValueError('log scale is not available when performance metric can be negative.')
+
+    return check_points, perf_by_time_step
+
+
+class ResultsVisualizer:  # TODO: Add tests
+    @staticmethod
+    def _set_plot_args(
+        ax: plt.Axes,
+        plot_setting_params: PlotSettingParams
+    ) -> None:
+        if plot_setting_params.xlim is not None:
+            ax.set_xlim(*plot_setting_params.xlim)
+        if plot_setting_params.ylim is not None:
+            ax.set_ylim(*plot_setting_params.ylim)
+
+        if plot_setting_params.xlabel is not None:
+            ax.set_xlabel(plot_setting_params.xlabel)
+        if plot_setting_params.ylabel is not None:
+            ax.set_ylabel(plot_setting_params.ylabel)
+
+        ax.set_xscale(plot_setting_params.xscale)
+        ax.set_yscale(plot_setting_params.yscale)
+        if plot_setting_params.xscale == 'log' or plot_setting_params.yscale == 'log':
+            ax.grid(True, which='minor', color='gray', linestyle=':')
+
+        ax.grid(True, which='major', color='black')
+
+        if plot_setting_params.legend:
+            ax.legend(loc=plot_setting_params.legend_loc)
+
+        if plot_setting_params.title is not None:
+            ax.set_title(plot_setting_params.title)
+        if plot_setting_params.show:
+            plt.show()
+
+    @staticmethod
+    def _plot_individual_perf_over_time(
+        ax: plt.Axes,
+        cum_times: np.ndarray,
+        results: np.ndarray,
+        worst_val: float,
+        plot_setting_params: PlotSettingParams,
+        label: Optional[str] = None,
+        color: Optional[str] = None,
+        *args: Any,
+        **kwargs: Any
+    ) -> None:
+        """
+        Plot the performance of the AutoPytorch over time.
```

Comment:
```suggestion
        Plot the incumbent performance of AutoPytorch over time.
```

Reply:
Actually, could you also add details on the difference between this and the `plot_perf_over_time`?

Reply:
This is an implicit method, so it is supposed to be referred only inside the class. In other words, whenever people see methods with preceding `_`, such methods should not be called from outside.

It is also in the autopep8 guideline:)

_single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.

Ref: [autopep8 Descriptive Naming Styles](https://www.python.org/dev/peps/pep-0008/#id34)

Reply:
Sure, but I meant for better documentation for the users, we can let them know that this is plotting performance for an individual step.